[RFC] Expand env (e.g. ~) in fname given to :profile

mhinz opened the pull request:
Eventually the fname is given to `fopen(3)` which doesn't expand things like `~`
and gives an error that it can't open the file.

References #2389.

fwalch commented:
I think this bug is also in Vim, right?

mhinz commented:
Well, I wouldn't say it's a bug per-se, but most people will find being able to use `~` pretty convenient.

justinmk commented:
cc @chrisbra

splinterofchaos commented:
LGTM :+1:

oni-link commented on src/nvim/ex_cmds2.c:
```diff
@@ -760,7 +760,7 @@ void ex_profile(exarg_T *eap)
   if (len == 5 && STRNCMP(eap->arg, "start", 5) == 0 && *e != NUL) {
     free(profile_fname);
-    profile_fname = vim_strsave(e);
+    profile_fname = expand_env_save(e);
```
Would be `expand_env_save_opt(c,true);` better? `c` should be one file name.

mhinz replied:
Right, that would be a better fit. Thanks!

mhinz replied:
Good call!

mhinz commented:
I replaced `vim_strsave(logfile)` by `expand_env_save_opt(logfile, true)` and refactored `:ex_profile()` while being at it. I hope the latter is not frowned upon.

justinmk commented on src/nvim/ex_cmds2.c:
```diff
@@ -744,40 +744,42 @@ void dbg_breakpoint(char_u *name, linenr_T lnum)
-  static proftime_T pause_time;
+  static proftime_T  pause_time;
+  int                len;
+  char_u            *arg;
```
why not move this to line 757? same for the others--move them to first usage if possible.

mhinz replied:
Oh, I thought that was conventional. Will change.

justinmk replied:
The legacyVim code follows that convention (and must, because of C89), but we do not.

mhinz replied:
It's not even C89 (:-P), but I get the point.

The `char_u` -> `char`/`uint8_t` change will make lots of code much cleaner. That will be a joyful day!

splinterofchaos commented:
Refactoring is great, but please separate cosmetic changes from functional ones. It makes it harder to see what changed.

mhinz commented:
Would you even accept a commit refactoring a single function?

mhinz committed:
Expand :profile's filename

vim_strsave() is replaced by expand_env_save_opt(), which expands ~ for
convenience:

    :profile start ~/.nvim/prof.log

Prior to this change you had to specify an absolute path.

splinterofchaos commented:
> Would you even accept a commit refactoring a single function?

Considering the amount of energy required to review a PR that refactors whole files/modules, personally, I would be more likely to review and want to push one with a small number of changes. Verifying large commits is both intellectually taxing and error prone, although a very necessary task.

> I also like having a clean version history opposed to series of commits of single line changes in one and the same file.

Right, that would be bad. A commit with functional changes can of course have some refactorings. But if the intent of the commit is not clear, or it's difficult to see how the functionality changed, it could confuse those reading through the history.

splinterofchaos commented on src/nvim/ex_cmds2.c:
```diff
@@ -744,40 +744,40 @@ void dbg_breakpoint(char_u *name, linenr_T lnum)
- */
+/// ":profile cmd args"
+///
+/// @param eap Pointer to the command arguments structure.
```
I think just the `:profile cmd args` is necessary here. Anyone wanting information on how this should work can run `:h :profile`, and the contents of `eap` is required reading for all `ex_cmds[2].c` functions.

mhinz replied:
Okeydokey

justinmk commented on src/nvim/ex_cmds2.c:
```diff
@@ -744,40 +744,38 @@ void dbg_breakpoint(char_u *name, linenr_T lnum)
+  const char *cmd = (char *)eap->arg;
 
-  if (len == 5 && STRNCMP(eap->arg, "start", 5) == 0 && *e != NUL) {
+  if (len == 5 && !strncmp(cmd, "start", 5) && *arg) {
```
Use `NUL` when it is semantically appropriate. Also I believe most code (even new code) that I have seen compares the explicit result of `strcmp` family of functions, for clarity. 

As far as `STRNCMP` vs `strncmp`, we will make a mass-action decision about this later.

justinmk replied:
Note however that we *do* tend to prefer `(foo)` and `(!foo)` instead of `(foo != NULL)` and `(foo == NULL)`, where `foo` is a pointer.

There are several style guide updates that are needed, I will be sure to address it soon.

justinmk commented on src/nvim/ex_cmds2.c:
```diff
@@ -744,40 +744,38 @@ void dbg_breakpoint(char_u *name, linenr_T lnum)
-  else if (STRCMP(eap->arg, "pause") == 0) {
-    if (do_profiling == PROF_YES)
+  } else if (do_profiling == PROF_NONE) {
+    emsg((char_u *)_("E750: First use \":profile start {fname}\""));
```
This doesn't gain use anything. When we remove char_u we will do a mass-migration of `emsg` calls. Until then, this change actually increases the work of removing char_u later.

justinmk commented on src/nvim/ex_cmds2.c:
```diff
@@ -744,40 +744,38 @@ void dbg_breakpoint(char_u *name, linenr_T lnum)
     set_vim_var_nr(VV_PROFILING, 1L);
-  } else if (do_profiling == PROF_NONE)
-    EMSG(_("E750: First use \":profile start {fname}\""));
-  else if (STRCMP(eap->arg, "pause") == 0) {
```
Maybe @elmart @aktau @splinterofchaos can offer an alternative opinion, but I believe we prefer the explicit comparison for strcmp family.

splinterofchaos replied:
I do prefer `strcmp(...) == 0`, but don't feel strongly about it. Actually, I'm tempted to suggest we add something like
```c
#define strequal(a,b) (strcmp(a,b) == 0)
```
to `strings.h`.

elmart replied:
I do agree to explicit comparison for `strcmp`. `strequal` would be a good addition, too. Maybe better defined as an inline function instead of a macro.

mhinz commented:
@justinmk Taking all that into account, I think there's no much use for that commit right now. I'll change the branch to only contain the actual fix.

justinmk commented:
Merged, thanks
